src/connector/src/parser/avro/util.rs

```diff
@@ -152,10 +164,152 @@ fn avro_type_mapping(schema: &Schema) -> ConnectorResult<DataType> {
+                            schema
+                        )
+                    }
+                }
+                None => {
+                    bail!("`map.handling.mode` not specified in ENCODE AVRO (...). Currently supported modes: `jsonb`")
+                }
+            }
+        }
+        Schema::Null | Schema::Fixed(_) | Schema::Uuid => {
             bail!("unsupported Avro type: {:?}", schema)
         }
     };
 
     Ok(data_type)
 }
+
+/// Check for [`avro_to_jsonb`]
+fn supported_avro_to_json_type(schema: &Schema) -> bool {
+    match schema {
+        Schema::Null | Schema::Boolean | Schema::Int | Schema::String => true,
+
+        Schema::Map(value_schema) | Schema::Array(value_schema) => {
+            supported_avro_to_json_type(value_schema)
+        }
+        Schema::Union(union_schema) => union_schema
+            .variants()
+            .iter()
+            .all(supported_avro_to_json_type),
+        Schema::Record(RecordSchema { fields, .. }) => fields
+            .iter()
+            .all(|f| supported_avro_to_json_type(&f.schema)),
+        Schema::Long
+        | Schema::Float
+        | Schema::Double
+        | Schema::Bytes
+        | Schema::Enum(_)
+        | Schema::Fixed(_)
+        | Schema::Decimal(_)
+        | Schema::Uuid
+        | Schema::Date
+        | Schema::TimeMillis
+        | Schema::TimeMicros
+        | Schema::TimestampMillis
+        | Schema::TimestampMicros
+        | Schema::LocalTimestampMillis
+        | Schema::LocalTimestampMicros
+        | Schema::Duration
+        | Schema::Ref { name: _ } => false,
+    }
+}
+
+pub(crate) fn avro_to_jsonb(
+    avro: &Value,
+    builder: &mut jsonbb::Builder,
+) -> crate::parser::AccessResult<()> {
+    match avro {
+        Value::Null => builder.add_null(),
+        Value::Boolean(b) => builder.add_bool(*b),
+        Value::Int(i) => builder.add_i64(*i as i64),
+        Value::String(s) => builder.add_string(s),
+        Value::Map(m) => {
+            builder.begin_object();
+            for (k, v) in m {
+                builder.add_string(k);
+                avro_to_jsonb(v, builder)?;
+            }
+            builder.end_object()
+        }
+        // same representation as map
+        Value::Record(r) => {
+            builder.begin_object();
+            for (k, v) in r {
+                builder.add_string(k);
+                avro_to_jsonb(v, builder)?;
+            }
+            builder.end_object()
+        }
+        Value::Array(a) => {
+            builder.begin_array();
+            for v in a {
+                avro_to_jsonb(v, builder)?;
+            }
+            builder.end_array()
+        }
+        Value::Union(_, v) => avro_to_jsonb(v, builder)?,
+
+        // TODO: figure out where the following encoding is reasonable before enabling them.
+        // jsonbb supports int64, but JSON spec does not allow it. How should we handle it?
+        // BTW, protobuf canonical JSON converts int64 to string.
```

Comment:
I tend to completely follow the Protobuf's canonical JSON conversion ([docs](https://protobuf.dev/programming-guides/proto3/#json)). I believe they have made the best trade-offs and we shouldn't waste time to discuss again.

Reply:
We can borrow ideas from it but it is hard to say that *conversion between avro and RisingWave jsonb just follows protobuf*. Anyways we will be targeting a native `map` type.

Reply:
> Anyways we will be targeting a native map type.

Sure.

> We can borrow ideas from it but it is hard to say that conversion between avro and RisingWave jsonb just follows protobuf.

I think the thing is, logically: `protobuf --> json --> jsonb`, where the `protobuf --> json` part should follow the Protobuf's canonical mapping, and `json -> jsonb` is much easier and less controversial than the former one.

Reply:
> they have made the best trade-offs

IMO one aspect they may not have considered is that whether it's easy to do data transformation on the JSON directly, specifically in Postgres. They should mainly care about ser/de only.

Reply:
> > they have made the best trade-offs
> 
> IMO one aspect they may not have considered is that whether it's easy to do data transformation on the JSON directly, specifically in Postgres. They should mainly care about ser/de only.

Yeah, I can understand your concern. In my mind, the best way to play with PG data types is to use the native `map` data type. As we have decided to do it in near future, I think for this PR we can ignore that requirement.

Reply:
> where the `protobuf --> json` part should follow the Protobuf's canonical mapping

And this PR is about avro. Avro spec also has `avro -> json` [here](https://avro.apache.org/docs/1.11.1/specification/#json-encoding) and it indeed made different decisions from protobuf on how to handle int64, nan and bytes ...

Reply:
> Avro spec also has `avro -> json` [here](https://avro.apache.org/docs/1.11.1/specification/#json-encoding)

Wait, I think that is avro's data serialization format (binary/json) instead of protobuf's "Type Mapping".

Reply:
> > where the `protobuf --> json` part should follow the Protobuf's canonical mapping
> 
> And this PR is about avro. Avro spec also has `avro -> json` [here](https://avro.apache.org/docs/1.11.1/specification/#json-encoding) and it indeed made different decisions from protobuf on how to handle int64, nan and bytes ...

Sorry for being stupid 🤡 

Sadly, the Avro spec seems not as detailed as Protobuf's. I tend to follow it as much as possible.

Regarding of `NaN`, to my best knowledge, it's not well-defined in the document. As mentioned in [AVRO-1290](https://issues.apache.org/jira/browse/AVRO-1290), 

> JSON specification does not say how to encode +infinity, -infinity or NaN. Consequently other language bindings of Avro do not honor infinity etc.

Perhaps we can handle it arbitrarily.

Reply:
> > Avro spec also has `avro -> json` [here](https://avro.apache.org/docs/1.11.1/specification/#json-encoding)
> 
> Wait, I think that is avro's data serialization format (binary/json) instead of protobuf's "Type Mapping".

They seem to share the same purpose - to standardize the conversion between JSON representation and the encoded data. 👀

Reply:
> In my mind, the best way to play with PG data types is to use the native map data type. As we have decided to do it in near future, I think for this PR we can ignore that requirement.

So my intention in this PR is to only allow very limited types (non controversial) in JSON now. And maybe enable more if requested by users in the future. The most urgent requesting this feature only needs `map<string>`. Do you want to allow more types in this PR? 🤔

Reply:
> > > Avro spec also has `avro -> json` [here](https://avro.apache.org/docs/1.11.1/specification/#json-encoding)
> > 
> > 
> > Wait, I think that is avro's data serialization format (binary/json) instead of protobuf's "Type Mapping".
> 
> They seem to share the same purpose - to standardize the conversion between JSON representation and the encoded data. 👀

Avro's JSON encoding includes schema information in it:
It is encoded as a JSON object with one **name/value pair whose name is the type’s name** and whose value is the recursively encoded value.

Reply:
> I tend to completely follow the Protobuf's canonical JSON conversion ([docs](https://protobuf.dev/programming-guides/proto3/#json)). I believe they have made the best trade-offs and we shouldn't waste time to discuss again (except the difference `jsonb` and real json).

Another off-topic point: Our FORMAT JSON sink doesn't follow it neither. 🤡

Reply:
> Avro's JSON encoding includes schema information in it:
> It is encoded as a JSON object with one **name/value pair whose name is the type’s name** and whose value is the recursively encoded value.

It is not schema information, but an essential part of the value. This is a special case made for union. The `name` part serves as the tag part of a tagged union, as I mentioned in https://github.com/risingwavelabs/risingwave/pull/16948#discussion_r1616638657

Reply:
> Another off-topic point: Our FORMAT JSON sink doesn't follow it neither.

We have discussed `protobuf -> json` and `avro -> json` above. The json encoder in sink likely followed a third convention: [`org.apache.kafka.connect.json.JsonConverter`](https://github.com/apache/kafka/blob/3.7.0/connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java#L603-L604) used by debezium. For completeness there is a fourth one: `to_jsonb` function defined by PostgreSQL.

Reply:
> > In my mind, the best way to play with PG data types is to use the native map data type. As we have decided to do it in near future, I think for this PR we can ignore that requirement.
> 
> So my intention in this PR is to only allow very limited types (non controversial) in JSON now. And maybe enable more if requested by users in the future. The most urgent requesting this feature only needs `map<string>`. Do you want to allow more types in this PR? 🤔

IMO yes, because it's a quite simple work regardless of our arguments here. 😄